### PR TITLE
MOD-7307: Add empty indexing validation

### PR DESCRIPTION
**Pull request by raz-mon**
**Describe the changes in the pull request**

Adds a validation that the user is not searching for empty values on a field that does not index them.

**Mark if applicable**

- [ ] This PR introduces API changes
- [ ] This PR introduces serialization changes

### CODE REVIEW

File: src/query.c [hunk 1]
```diff
@@ -1656,6 +1705,18 @@ int QueryNode_CheckIsValid(QueryNode *n, IndexSpec *spec, RSSearchOptions *opts,
...
 }
 
+// Checks whether query nodes are valid
+// Currently Phrase nodes are checked whether slop/inorder are allowed
+int QAST_CheckIsValid(QueryAST *q, IndexSpec *spec, RSSearchOptions *opts, QueryError *status) {
+  if (!q || !q->root ||
+      !(spec->flags & Index_HasNonEmpty) &&
+      (!isSpecJson(spec) || !(spec->flags & Index_HasUndefinedOrder))
+  ) {
+    return REDISMODULE_OK;
```
Comment (oshadmi) [inline src/query.c:1715]: Might it skip the Tag node check?

Comment (raz-mon) [reply src/query.c:1715]: Notice the `!(spec->flags & Index_HasNonEmpty)` condition

**Comment by nafraf** [comment]
Is this expected?
```sh
127.0.0.1:6379> FT.CREATE idx schema t1 text indexempty t2 text t3 text indexempty
OK
127.0.0.1:6379> FT.SEARCH idx '@t2|t1:("")' dialect 2
1) (integer) 0
(0.77s)
127.0.0.1:6379> FT.SEARCH idx '@t2:("") | @t1:("")' dialect 2
(error) Empty value for field that does not index empty values
(1.55s)
```

File: tests/pytests/test_empty.py [hunk 2]
```diff
@@ -3,6 +3,62 @@
...
+        # an error
+        res = conn.execute_command('FT.SEARCH', 'idx', '')
+        env.assertEqual(res, EMPTY_RESULT)
+        res = conn.execute_command('FT.SEARCH', 'idx', '""')
+        env.assertEqual(res, EMPTY_RESULT)
+        res = conn.execute_command('FT.SEARCH', 'idx', "''")
+        env.assertEqual(res, EMPTY_RESULT)
+
+        # Any query containing "@tag:{''}" or "@text:''" should throw an error
+        for query in ['@tag:{""}', "@tag:{''}", '@text:""', "@text:''"]:
```
Comment (nafraf) [inline tests/pytests/test_empty.py:30]: Bug? Consider testing with two tag fields:
```sh
127.0.0.1:6379> FT.CREATE idx schema t1 tag indexempty t2 tag
OK
127.0.0.1:6379> FT.SEARCH idx '@t1:("")' dialect 2
(error) Empty value for field that does not index empty values
127.0.0.1:6379> FT.SEARCH idx '@t2:("")' dialect 2
(error) Empty value for field that does not index empty values
```

Comment (raz-mon) [reply tests/pytests/test_empty.py:30]: Right!
Corrected this back to return `[0]`, as before, in order not to break (this should be an error message..).

File: src/query.c [hunk 3]
```diff
@@ -1605,31 +1605,90 @@ int QueryNode_EvalParams(dict *params, QueryNode *n, QueryError *status) {
...
+    array_free(fields);
+  }
+  return empty_text;
+}
+
+// If the token is of an empty string, and the searched field doesn't index
+// empty strings, we should return an error
+static inline bool QueryNode_ValidateToken(QueryNode *n, IndexSpec *spec, RSSearchOptions *opts, QueryError *status) {
+  if (n->tn.len == 0 && n->tn.str && !strcmp(n->tn.str, "") && !QueryNode_DoesIndexEmpty(n, spec, opts)) {
+    QueryError_SetError(status, QUERY_ESYNTAX, "Empty value for field that does not index empty values");
```
Comment (oshadmi) [inline src/query.c:1649]: We used to have a more constructive error message such as in commit 10261d2f9
```
QueryError_SetErrorFmt(q->status, QUERY_ENOINDEX,
+                              "Field `%s` should enable `%s` in the index SCHEMA in order to support empty values",
+                              IndexSpec_GetFieldNameByBit(q->sctx->spec, qn->opts.fieldMask),
+                              SPEC_INDEXEMPTY_STR);
```

Comment (raz-mon) [reply src/query.c]: 👍 
Changed to a positive message: `QueryError_SetErrorFmt(status, QUERY_ESYNTAX, "Use `%s` in field creation in order to index and query for empty strings", SPEC_INDEXEMPTY_STR);`

**Review by oshadmi — APPROVED** [review]

**Pull request merged**